modules/python/Makefile.env

```diff
@@ -1,41 +1,36 @@
 PY_PIPENV := $(shell which pipenv 2>/dev/null)
-PY_PIPFILE_EXISTS:=$(shell echo "$$(test -f Pipfile; echo $$?)")
-
 ifneq ($(PY_PIPENV),)
-	PY_CURRENT_VENV:=$(shell python -c 'from __future__ import print_function; import sys; print(sys.prefix if hasattr(sys, "real_prefix") or (hasattr(sys, "base_prefix") and sys.base_prefix != sys.prefix) else "", end="")')
-	ifeq ($(PY_CURRENT_VENV),)
-		PY_VENV:=.venv
+	# Unless WORKON_HOME is set, assume we're going to create ".venv" next to the Pipfile.
+	ifeq ($(WORKON_HOME),)
+		PIPENV := PIPENV_VENV_IN_PROJECT=1 $(PY_PIPENV)
 	else
-		PY_VENV:=$(PY_CURRENT_VENV)
+		PIPENV := PIPENV_VENV_IN_PROJECT= WORKON_HOME='$(WORKON_HOME)' $(PY_PIPENV)
 	endif
+	WITH_PIPENV := $(PIPENV) run
 
-	PY_VENV_WORKDIR?=.
-
-	ifeq ($(PY_PIPFILE_EXISTS),0)
-		ifeq ($(PY_VENV_WORKDIR),.)
-			PIPENV:=PIPENV_VENV_IN_PROJECT=1 $(PY_PIPENV)
-		else
-			PIPENV:=PIPENV_VENV_IN_PROJECT= WORKON_HOME='$(shell realpath $(PY_VENV_WORKDIR))' $(PY_PIPENV)
-			# if the venv doesn't exist then this will fail and PY_VENV will be empty
-			PY_VENV:=$(shell $(PIPENV) --venv 2> /dev/null)
-			ifneq ($(.SHELLSTATUS),0)
-				# means pipenv failed, so it doesn't exist yet. Need to cause a make env
-				_:=$(shell $(value PIPENV) install --dev --deploy)
+	# Get the directory where the Pipfile is.
+	PIPENV_HOME := $(shell $(PIPENV) --where 2> /dev/null)
 
-				PY_VENV:=$(shell $(PIPENV) --venv 2> /dev/null)
-			endif
+	ifeq ($(WORKON_HOME),)
+		# Again, assume we're going to create ".venv" next to the Pipfile.
+		PY_VENV := $(PIPENV_HOME)/.venv
+	else
+		# If the venv doesn't exist yet, then this will fail and PY_VENV will be empty
+		PY_VENV := $(shell $(PIPENV) --venv 2> /dev/null)
+		ifneq ($(.SHELLSTATUS),0)
+			# Create the venv, then check for it's location again.
+			_ := $(shell $(WITH_PIPENV) install --dev --deploy)
+			PY_VENV := $(shell $(PIPENV) --venv 2> /dev/null)
```

Comment:
Now that I'm thinking about it...do we need this logic that installs the pipenv immediately, outside of a target when WORKON_HOME is set?

Reply:
While we're at it, do we need to handle `WORKON_HOME` at all?

Reply:
I was thinking of using it for the build-harness magic venv, but we certainly don't use it anywhere on our projects right now.

Reply:
Maybe we just remove it?

Reply:
Wait, yes, that's how Jenkins works.

Reply:
Does it? I just checked a build and didn't see WORKON_HOME set

Reply:
How about something like this?

```makefile
PY_PIPENV := $(shell which pipenv 2>/dev/null)
ifneq ($(PY_PIPENV),)
	# Unless WORKON_HOME is set, assume we're going to create ".venv" next to the Pipfile.
	ifeq ($(WORKON_HOME),)
		PIPENV := PIPENV_VENV_IN_PROJECT=1 $(PY_PIPENV)
		PY_VENV := $(CURDIR)/.venv
	else
		PIPENV := PIPENV_VENV_IN_PROJECT= WORKON_HOME='$(WORKON_HOME)' $(PY_PIPENV)
		PY_VENV := $(WORKON_HOME)/.venv
	endif
	WITH_PIPENV := $(PIPENV) run

	# Get the directory where the Pipfile is.
	PIPENV_HOME := $(shell $(PIPENV) --where 2> /dev/null)
	ifeq($(PIPENV_HOME),)
		PIPENV_HOME := $(CURDIR)
	endif
endif
PY_VENV ?= $(CURDIR)/.venv
PIPENV_HOME ?= $(CURDIR)
```

Reply:
I've just learned that targets can be defined or not by conditions:
```Makefile
ifneq ($(PIPENV_HOME),)

$(PIPENV_HOME)/Pipfile.lock: $(PIPENV_HOME)/Pipfile
	@$(MAKE) pipenv/lock

endif
```

Reply:
Or we could use `$(call assert-set,PIPENV_HOME)`.

Reply:
Where would you call that?

Reply:
I think it's ok to assume / default to `PIPENV_HOME?=$(CURDIR)` as long as the targets being run depend on $(PIPENV_HOME)/Pipfile existing

Reply:
```makefile
PY_PIPENV := $(shell which pipenv 2>/dev/null)
ifneq ($(PY_PIPENV),)
	...(as above)...

	# Get the directory where the Pipfile is.
	PIPENV_HOME := $(shell $(PIPENV) --where 2> /dev/null)
endif

PIPENV_HOME?=

ifeq ($(PIPENV_HOME),)
	PY_VENV := $(CURDIR)/.venv
	PIPENV_HOME := $(CURDIR)
endif

$(PIPENV_HOME)/Pipfile:
	@echo "ERROR: You must create a Pipfile!"
	@exit 1
```

gives me

```
> $ make pipenv
Found python: /Users/charliemathews/.pyenv/shims/python
Found pipenv: /Users/charliemathews/.pyenv/shims/pipenv
ERROR: You must create a Pipfile!
make: *** [/Users/charliemathews/build-harness-test/Pipfile] Error 1
```

Creating a `Pipfile` causes it to behave as expected.

Reply:
```
> $ make requirements.txt
ERROR: You must create a Pipfile!
make: *** [/Users/charliemathews/build-harness-test/Pipfile] Error 1
```

Reply:
@alphachai I'm going to resolve this thread because I just pushed a bunch of changes. Now if there's no Pipfile, `PY_VENV` won't get set. Then the pipenv targets check to make sure the variables they need are set.